### PR TITLE
Corrected: Fix for "Too many files open" bug

### PR DESCRIPTION
**Pull request by raygesualdo**
The error raised in #3 occurs when there's a lot of files that `split-guide` has to read. I found the number for small files to be somewhere between 1000 and 10,000 (the number changes based on system resources available and file size). To solve this, I wrote a utility function that takes all of the files that need to be opened, chunks them out into smaller groups, runs `Promise.all` on each of the chunks serially, then returns the whole set. I've tested this with an exercise set of 10,000 files and it works great. I tried 100,000 files and Node completely crashed on me with an out of memory exception, but hopefully no one will ever use this with that many files :)

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -49,10 +50,23 @@ function splitGuide({
     return pify(fs.readFile)(file, 'utf8')
       .then(contents => ({file, contents}))
   }
+  
+  function promiseChunker(remainingArrayElements, chunkSize) {
```
Comment (kentcdodds) [inline src/index.js]: Hey, this is a pretty nifty function!

File: src/index.js [hunk 2]
```diff
@@ -26,6 +26,7 @@ function splitGuide({
     .then(readAllFilesAsPromise)
     .then(createNewFileContents)
     .then(saveFiles)
+    .then(reduceFilenames)
```
Comment (kentcdodds) [inline src/index.js]: What was the purpose of adding this? I don't think I mind it, but I'm curious why you added it :)

File: src/index.js [hunk 3]
```diff
@@ -82,17 +96,23 @@ function splitGuide({
...
     const relativeDestination = path.relative(templatesDir, file)
     const workshopDestination = path.resolve(exercisesDir, relativeDestination)
     const finalDestination = path.resolve(exercisesFinalDir, relativeDestination)
     return [
-      workshopContents ? saveFile(workshopDestination, workshopContents) : null,
-      finalContents ? saveFile(finalDestination, finalContents) : null,
-    ].filter(Boolean) // filter out the files that weren't saved
+      workshopContents ? [saveFile, workshopDestination, workshopContents] : null,
+      finalContents ? [saveFile, finalDestination, finalContents] : null,
+    ].filter(p => !!p) // filter out the files that weren't saved
```
Comment (kentcdodds) [inline src/index.js]: Any reason this was changed from `].filter(Boolean)`?

**Review by kentcdodds — APPROVED** [review]
Looks pretty awesome. Just a few questions, but I don't think we need to change anything!

File: src/index.js [hunk 4]
```diff
@@ -49,10 +50,23 @@ function splitGuide({
     return pify(fs.readFile)(file, 'utf8')
       .then(contents => ({file, contents}))
   }
+  
+  function promiseChunker(remainingArrayElements, chunkSize) {
+    if (remainingArrayElements.length === 0) return Promise.resolve([])
```
Comment (kentcdodds) [inline src/index.js]: Oh, linting is failing because I prefer to put if statements in blocks.

**Review by kentcdodds — CHANGES_REQUESTED** [review]
Let's fix linting really quick, then I think we're good :)

File: src/index.js (hunk 1 above)
Comment (raygesualdo) [reply src/index.js]: Thanks! It took me a bit to come up with what's here, but I'm pleased with how it turned out.

File: src/index.js (hunk 4 above)
Comment (raygesualdo) [reply src/index.js]: Oops, I'll get that taken care of.

File: src/index.js (hunk 3 above)
Comment (raygesualdo) [reply src/index.js]: No. I think that was an error when I updated my branch with the latest from your repo. I'll revert it.

File: src/index.js (hunk 1 above)
Comment (kentcdodds) [reply src/index.js]: Actually, I know that sindre sorhus came up with a [bunch of promise related modules](https://github.com/sindresorhus/promise-fun) recently. Maybe he's got a utility like this...

Comment (kentcdodds) [reply src/index.js]: Maybe this one? https://github.com/sindresorhus/p-limit

Comment (raygesualdo) [reply src/index.js]: That would also work. Would you like me to swap it out?

Comment (kentcdodds) [reply src/index.js]: Yeah, it's good to use well tested modules I think. Sorry!

**Comment by raygesualdo** [comment]
Made the changes you requested. The `p-limit` package was a good call. Cleaned up the PR considerably.

**Review by kentcdodds — APPROVED** [review]
Super!

**Pull request merged**